authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerHiveAuthorizationPlugin.java

```diff
@@ -0,0 +1,840 @@
+    if (list.getListSize() == 0) {
+      LOG.warn("The group({}) is not exist in the Ranger!", group);
+      return Boolean.FALSE;
+    }
+    return rangerClient.deleteGroup(list.getList().get(0).getId());
+  }
+
+  @Override
+  public Boolean onGroupAcquired(Group group) {
+    VXGroupList vxGroupList = rangerClient.searchGroup(ImmutableMap.of("name", group.name()));
+    if (vxGroupList.getListSize() == 0) {
+      LOG.warn("The group({}) is not exist in the Ranger!", group);
+      return Boolean.FALSE;
+    }
+    return Boolean.TRUE;
+  }
+
+  /**
+   * Add the securable object's privilege to the policy. <br>
+   * 1. Find the policy base the securable object. <br>
+   * 2. If the policy is exist and have same privilege, because support idempotent operation, so
+   * return true. <br>
+   * 3. If the policy is exist but have different privilege, also return true. Because one Ranger
+   * policy maybe contain multiple Gravitino securable object <br>
+   * 4. If the policy is not exist, then create a new policy. <br>
+   */
+  private boolean doAddSecurableObject(String roleName, RoleChange.AddSecurableObject change) {
+    RangerPolicy policy = findManagedPolicy(change.getSecurableObject());
+
+    if (policy != null) {
+      Set<String> policyPrivileges =
+          policy.getPolicyItems().stream()
+              .flatMap(policyItem -> policyItem.getAccesses().stream())
+              .map(RangerPolicy.RangerPolicyItemAccess::getType)
+              .collect(Collectors.toSet());
+      Set<String> newPrivileges =
+          change.getSecurableObject().privileges().stream()
+              .filter(Objects::nonNull)
+              .flatMap(privilege -> translatePrivilege(privilege.name()).stream())
+              .filter(Objects::nonNull)
+              .collect(Collectors.toSet());
+      if (policyPrivileges.containsAll(newPrivileges)) {
+        LOG.info(
+            "The privilege({}) already added to Ranger policy({})!",
+            policy.getName(),
+            change.getSecurableObject().fullName());
+        // If exist policy and have same privilege then directly return true, because support
+        // idempotent operation.
+        return true;
+      }
+    } else {
+      policy = new RangerPolicy();
+      policy.setService(rangerServiceName);
+      policy.setName(formatPolicyName(roleName, change.getSecurableObject().fullName()));
+      policy.setPolicyLabels(Lists.newArrayList(MANAGED_BY_GRAVITINO));
+
+      List<String> metadataObjectNamespaces =
+          Lists.newArrayList(
+              SecurableObjects.DOT_SPLITTER.splitToList(change.getSecurableObject().fullName()));
+      if (metadataObjectNamespaces.size() > 4) {
+        // The max level of the securable object is `catalog.db.table.column`
+        throw new RuntimeException("The securable object than 4");
+      }
+      metadataObjectNamespaces.remove(0); // remove `catalog`
```

Comment:
Why do we not need to set the catalog in `PolicyResource`, how do we uniquely identify the resource during authorization?

Reply:
Ranger Hive only has `database`, `table`, and `column`, It doesn't have `catalog`.